tools/single_use/generate_password_blacklist.php

```diff
@@ -0,0 +1,127 @@
+a table in the database. This "blacklist" will be analyzed
+upon password changes. If a user tries to use a password that appears in the
+list of hashes, it will be rejected.
+
+This script will also expire all current passwords for LORIS users so that
+exposed passwords cannot be used again.
+
+Instructions:
+- Run `2018-10-22-CreatePasswordBlacklistTable.sql`. This SQL file will
+create the table containing the blacklist entries.
+- Run this script to populate the table with the plaintext passwords in 
+the `history` table.
+
+The plaintext passwords appeared due to a bug in LORIS. As these passwords
+have potentially been viewed by administrators or could be read in the event
+of a data breach, we have chosen to reject them in this way.
+
+
+
+INFO;
+
+echo $info;
+$answers = array('y', 'N');
+$defaultAnswer = 'N';
+echo writeQuestion('Continue?', $answers);
+if (readAnswer($answers, $defaultAnswer) !== 'y') {
+    exit;
+}
+
+/* This script will write the password blacklist to a table in the DB. Fail if
+ * the user has not applied the SQL patch yet.
+ */
+try {
+    $sql = "SELECT 1 FROM password_blacklist LIMIT 1;";
+    $DB->pselectOne($sql, array());
+} catch (\DatabaseException $e) {
+    die(
+        '`password_blacklist` table not found. Please run SQL patch ' .
+        '`2018-10-22-CreatePasswordBlacklistTable.sql` or the appropriate ' .
+        'release patch.' .
+        PHP_EOL
+    );
+}
+
+// Query DB for burned passwords. The bug caused passwords to be stored on 
+// update so we'll limit the query to that. Additionally we will filter out
+// results from the `new` data that are password hashes i.e. those that begin
+// with the $ character.
+$sql = "select new as password from history where col = 'Password_hash' AND type='U' AND new NOT LIKE '\$%';";
+$result = $DB->pselectCol($sql, array());
+
+// Hash all the passwords.
+$password_hashes = array_map(
+    'hash_password',
+    array_values($result)
+);
+// Add hashes to blacklist table.
+foreach ($password_hashes as $password_hash) {
+    $DB->insert('password_blacklist', array('password_hash' => $password_hash));
+}
+// Expire all passwords. Must first fetch all user IDs from database as the
+// LORIS DB update function requires a WHERE statement.
+$result = $DB->pselectCol('select ID from users where ID > 0', array());
+foreach ($result as $id) {
+    $DB->update(
+        'users', 
+        array('Password_expiry' => '1990-01-01'), 
+        array('ID' => $id)
+    );
+}
+
+// Delete passwords from history table
+$DB->delete(
+    'history', array(
+        'col' => 'Password_hash',
+        'tbl' => 'users'
+    )
+);
+$DB->delete(
+    'history', array(
+        'col' => 'Password_md5',
+        'tbl' => 'users'
+    )
+);
+
+// Function exists because array_map doesn't support multiple arguments.
+function hash_password($password): string
+{
+    return password_hash($password, PASSWORD_DEFAULT);
```

Comment:
For the purposes of a blacklist, I don't see how this is going to work. (I also don't see where the blacklist is being compared against in the PR to check how it's supposed to be working..)

password_hash is not going to be reproducible and will change with each run, and the algorithm changes with different PHP versions by design. It's also intended to be slow, while you want the blacklist checking to be relatively fast.

Reply:
The functionality is implemented in #3962.

>password_hash is not going to be reproducible and will change with each run
I don't know what you mean here. Password hashing should be reproducible and I'm using the built-in functions that we rely on to create passwords in the first place. 

> the algorithm changes with different PHP versions by design
We can use `password_verify` which should always correspond to `password_hash`.

> It's also intended to be slow, while you want the blacklist checking to be relatively fast
This PR is a single_use script so the speed shouldn't really be a factor.
If speed proves to be an issue in blacklist checking in the other PR, we can do a `password_hash` once on the input and do a string `===` check on blacklist values coming from the DB. This would be faster than doing `password_verify` in a for-each loop (current implementation).

Reply:
We can't do a password hash on the input and a string comparison, we'll never be able to do that because by design password_hash will generate a different hash each time it's run. This isn't verifying passwords, this is checking if a password is blacklisted. Using password_hash is not scalable. In order to check whether or not a password is blacklisted, you need to run a (intentionally) slow and changing hashing algorithm against every single blacklisted password.